source/extensions/filters/http/cache/hazelcast_http_cache/hazelcast_http_cache.cc

```diff
@@ -0,0 +1,178 @@
+            max_body_size_);
+  HazelcastClusterAccessor& cluster_accessor = static_cast<HazelcastClusterAccessor&>(*accessor_);
+  ENVOY_LOG(info,
+            "Cache statistics can be observed on Hazelcast Management Center"
+            " from the map named {}.",
+            unified_ ? cluster_accessor.responseMapName() : cluster_accessor.headerMapName());
+}
+
+void HazelcastHttpCache::shutdown(bool destroy) {
+  if (!accessor_) {
+    ENVOY_LOG(warn, "Cache is already offline.");
+    return;
+  }
+  if (accessor_->isRunning()) {
+    ENVOY_LOG(info, "Shutting down Hazelcast connection...");
+    accessor_->disconnect();
+    ENVOY_LOG(info, "Cache is offline now.");
+  } else {
+    ENVOY_LOG(warn, "Hazelcast client is already disconnected.");
+  }
+  if (destroy) {
+    accessor_.reset();
+  }
+}
+
+HazelcastHttpCache::~HazelcastHttpCache() { shutdown(true); }
+
+LookupContextPtr HazelcastHttpCache::makeLookupContext(LookupRequest&& request) {
+  if (unified_) {
+    return std::make_unique<UnifiedLookupContext>(*this, std::move(request));
+  } else {
+    return std::make_unique<DividedLookupContext>(*this, std::move(request));
+  }
+}
+
+InsertContextPtr HazelcastHttpCache::makeInsertContext(LookupContextPtr&& lookup_context) {
+  ASSERT(lookup_context != nullptr);
+  if (unified_) {
+    return std::make_unique<UnifiedInsertContext>(*lookup_context, *this);
+  } else {
+    return std::make_unique<DividedInsertContext>(*lookup_context, *this);
+  }
+}
+
+// TODO(enozcan): Implement when it's ready on the filter side.
+//  Depending on the filter's implementation, the cached entry's
+//  variant_key_ must be updated as well. Also, if vary headers
+//  change then the hash key of the response will change and
+//  updating only header map will not be enough in this case.
+void HazelcastHttpCache::updateHeaders(LookupContextPtr&&, Http::ResponseHeaderMapPtr&&) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+constexpr absl::string_view HazelcastCacheName = "envoy.extensions.http.cache.hazelcast";
+
+// Cluster wide cache statistics should be observed on Hazelcast Management Center.
+// They are not stored locally.
+CacheInfo HazelcastHttpCache::cacheInfo() const {
+  CacheInfo cache_info;
+  cache_info.name_ = HazelcastCacheName;
+  cache_info.supports_range_requests_ = true;
+  return cache_info;
+}
+
+std::string HazelcastHttpCacheFactory::name() const { return std::string(HazelcastCacheName); }
+
+ProtobufTypes::MessagePtr HazelcastHttpCacheFactory::createEmptyConfigProto() {
+  return std::make_unique<HazelcastHttpCacheConfig>();
+}
+
+HttpCache& HazelcastHttpCacheFactory::getCache(
+    const envoy::extensions::filters::http::cache::v3alpha::CacheConfig& config) {
+  if (!cache_) {
+    HazelcastHttpCacheConfig hz_cache_config;
+    MessageUtil::unpackTo(config.typed_config(), hz_cache_config);
+    cache_ = std::make_unique<HazelcastHttpCache>(hz_cache_config);
```

Comment:
This isn't threadsafe. Envoy creates exactly one HazelcastHttpCacheFactory, but getCache is called multiple times, so more than one thread may try to assign to cache_ at the same time. Making cache_ a thread-local static might be the simplest efficient fix.

Reply:
tbh I could not manage to make it thread safe with static thread_local unique ptr. However I'd like to hear your opinion on creating the `cache_` via call_once or double-checked locking.

Reply:
First, two bits of background, because I have a hard time keeping some of this straight:
1. There is one HazelcastHttpCacheFactory, created and owned by Registry::RegisterFactory<HazelcastHttpCacheFactory, HttpCacheFactory>.
2. For every new connection, HazelcastHttpCacheFactory::getCache will be called on whichever worker thread is serving that connection, so getCache must expect concurrent calls.

HazelcastHttpCacheFactory::getCache returns pointers to the same HazelcastHttpCache. Is that the intent? If so, then HazelcastHttpCache must be threadsafe (and so documented), and I suggest creating cache_ in HazelcastHttpCacheFactory's constructor.

If you want to have one HazelcastHttpCache per worker thread, then a static thread_local unique_ptr would be a likely choice, and would make getCache threadsafe.

If you want to have one HazelcastHttpCache per connection, then getCache should create and return one, except the getCache interface makes lifetime difficult. If that's the intent, we'll need to change the interface to better support it.

Reply:
> HazelcastHttpCacheFactory::getCache returns pointers to the same HazelcastHttpCache. Is that the intent? 

Yes indeed. It uses thread safe APIs of Hazelcast. Other than the `connect` and `shutdown` methods, the cache itself is actually thread safe. I will modify both accordingly and document about it. Also I think it's better to use only one client (and so cache) for now and make them per-worker or per-connection or whatever later on in case of performance issues.

> and I suggest creating cache_ in HazelcastHttpCacheFactory's constructor.

But in that case it would not be able to connect until the first getCache call as it needs typed_config from CacheConfig. So a similar logic to ptr check would be required.

Reply:
Different calls to getCache can have different CacheConfigs, due to configuration changing over time, or due to different listeners instantiating the CacheFilter with different configurations. Does Hazelcast need global and/or static config?

Reply:
Well, I afraid I did not take this dynamic configuration change into consideration..
HazelcastClient must be configured before it starts. It takes a few seconds to be started (and terminated) and hence it's not applicable to do that for each getCache call. A possible 
solution might be using only one HazelcastClient configured with the initial config values. So getCache will simply ignore the client connection configurations but will create a new cache according to the typed config with the existing HazelcastClient (created with the very first getCache call.). This could be somehow acceptable. Because once a client connects, it can keep its connection alive regardless of the changes happened afterwards. However, creating a new cache per call sounds really poor to me. In that case I should probably get rid of start, shutdown, etc. stuff as much as possible and minimize the cost of cache initialization.

Reply:
@jmarantz, can you advise on how an extension should get static config?

Reply:
@toddmgreer, I've been thinking on this and decided to configure Hazelcast statically. However, I could not come up with a solution for other configurations (i.e. HazelcastHttpCacheConfig::unified/body_partition_size/app_prefix). 

I tried to create a single cache and expose these configurations to lookup and insertion contexts. That would solve the dynamic configuration issue for the cache since contexts can be arranged to operate with these values. But passing them to contexts seems not possible using a single cache for all threads.

I also considered to create a cache per getCache call with the passed CacheConfig's non-static configurations (unified, body_partition_size and app_prefix). But since we're returning a reference to the cache here, I got stuck again.

Let me ask this way: How the implementation would be like if SimpleHttpCache needed a config value from typed_config during lookup and/or insertion? If caches are supposed to be independent from this type of configuration, what's the point of typed_config then?

Thanks.

Reply:
@toddmgreer, a kindly reminder as I will change the impl. according to your answer.

Reply:
@enozcan, I apologize for not responding to your June 17th comment--I somehow didn't see it until just now.

While some types of config (such as those that HazelCast requires at startup) are naturally static, there are others that could be different for each request. Suppose SimpleHttpCache had a configuration option that specified a list of Content-Types not to insert into cache, for the purpose of not wasting cache space on content-types known (for that deployment) to get poor hit rates. SimpleHttpCacheFactory could create a request-specific HttpCache implementation that rejects those insertions, and otherwise inserts into a global hash map. Requests that hit differently configured filters might have different Content-Type blocklists, and that's fine.

I don't know the best way to handle global config. I intend to find out, as it's probably needed by all cache implementations (e.g. global limits for SimpleHttpCacheFactory, where to send RPCs for anything distributed). Unfortunately, there's some other work I have to get done first. If you're able to figure it out first, I'll try to be more prompt with discussions and PR reviews.